### DISCUSSION TITLE
Battery usage

### DISCUSSION THREAD

**Discussion by simonschiller** [MobileNativeFoundation/discussions#120] [General]
Hi everybody 👋 

I'd be curious to know if and how other folks are monitoring and measuring the battery usage of their apps in production. Are you monitoring it at all? Do you use proxy metrics to estimate how much battery your app is using? Are you collecting other metrics that relate to battery usage?

We just started the process of creating some observability for battery usage at Spotify. Our idea is to measure the current battery level on app start and periodically after that (only when the app is running anyways). We'll most likely filter out measurements where the device is currently charging, but that's how we plan on getting some high-level insights into battery consumption data. The data for a single user depends on many factors (like other running apps, device age, temperature, ...) and would not be very useful by itself. But on a large enough scale with enough users, we think the aggregated data could be useable as a metric in experiments or to detect general regressions. It's still up in the data will actually be useable though.

**Comment by Koshub** [comment]
Just a side note. It is interesting if Apple will allow to track it using statistics from millions of users. IMO, since the battery value has the OS dependency, it partially reflects how the OS is managing the battery (also, considering this API is available in iOS >= 3.0, it is a possible candidate for the privacy API review).

**Comment by amanjeetsingh150** [comment]
Hey, @simonschiller are you trying to capture this data for Android or iOS? We recently worked on battery drain in our organization and have optimized for wake-up alarms. We have experienced that the way to capture the battery level of the device is error-prone because it's a system-level metric and can give you wrong leads.

**Comment by simonschiller** [comment]
Hi @amanjeetsingh150! The plan is to do it for both platforms, starting with Android. You're right that battery drain by itself is not a very good metric if you look at for a single user, as there are a lot of factors other than our app (other apps, temperature, battery age, ...).

Our idea is that by the [law of large numbers](https://en.wikipedia.org/wiki/Law_of_large_numbers), the data should get more reliable once we collect it for millions of users, as the only thing all of them are going to have in common is our app. But we also don't know yet if this hypothesis is true and the data is actually going to be useable, or if the variance will still be to high to actually deduct meaningful information.

One other thing we're looking at is capturing additional information, so interpreting the data gets easier. For example, the device screen is one of the biggest battery consumers, so measuring if the screen is turned on or off could get us more info. We also have to be careful here, as observing too many things will also drain the battery - and we definitely don't want to do that.

**Reply by amanjeetsingh150** [reply]
Hey @simonschiller, Yup makes sense. According to the experience we had with battery drain, we didn't go all out on the battery holistically with battery level as we experienced too much noise on that as you already suggested. We solved for one of our major draining contributors taking it as a proxy metric to get an understanding of the whole battery domain.

Further, we can enlist all the battery drain contributors which are applicable on your app and think about capturing all of the related metrics for example wakeup alarms can be captured through [HealthStats](https://developer.android.com/reference/android/os/health/HealthStats) API, CPU usage can be mapped from [rusage](https://man7.org/linux/man-pages/man2/getrusage.2.html) API or Proc file of Linux through CPU time metric, etc.

I would also recommend you to go through [Battery Metrics](https://github.com/facebookincubator/Battery-Metrics). We can also model battery drain through [power profile](https://source.android.com/devices/tech/power/values) in android and calculate the exact maH and hence percentage (good approximation) to get battery drained done by the feature which can help us to reduce noise involved due to system-level metrics.  

One of the good talks I found while working on this can be found here: https://www.youtube.com/watch?v=daqI7Pr9zrE

Regarding the overhead, I would recommend capturing this telemetry conditionally on the basis of some cohort. For example, only in beta or some minimal percentage with rotated users, or users rolling out with some major feature.

Hope my answer helped 😄

**Comment by simonschiller** [comment]
Those are some really interesting and helpful pointers, thanks a lot! Definitely more advanced than what we've had in mind for our MVP - but now we have some of avenues to go down for future improvements.

I'll also make sure to update this discussion as soon as we have some learnings 😄
